### DISCUSSION TITLE
xunit v3 release plans and runner integration

### DISCUSSION THREAD

**Discussion by remcomulder** [xunit/xunit#2594] [Question] [answered]
Hi, it's Remco here from the NCrunch project.

We've been watching the progress on xunit v3 with interest and are trying to make long term plans on how best to handle our integration with it.

My assumption is that v3 will include significant API changes over v2 (please do tell me if this is incorrect), as such, I'm expecting that we'll need to build a new adapter to integrate with it.

Is there any plan in place involving a ballpark release date for v3?  Will there be much warning before it is released?  Ideally we'd be hoping for a few weeks to plumb something in on a release-ready build before it gets switched to stable on Nuget.

**Comment by bradwilson** [comment, marked as answer]
I'm sorry to say I still don't have a target release date. I took December off from working on it, but I will be getting back to it in January.

The alpha => beta => RTM will take significantly longer than a few weeks, and in particular in the alpha stage we would be interested in getting feedback on whether APIs meet needs. If you have proposals for APIs you would like to see, you don't have to wait: open new issue(s) as appropriate and I'll triage them.

**Reply by MartyIX** [reply]
Microsoft uses xUnit in the dotnet/runtime and in the dotnet/roslyn repos (just two I know). Aren't these guys good candidates to provide feedback for xUnit 3?

I mean I guess a lot of people can share their concerns and their feedback but these big projects seem like ideal testbeds really.

**Reply by bradwilson** [reply]
Yes, they are good candidates, but they will not use it until it supports the runner inside Visual Studio (which is waiting on #2337 plus an update to the VSTest adapter once it's available). Whether they will choose to do so is really up to them; I am not on that team (and not associated with Microsoft at all), so that's something that they would have to choose to do.

**Comment by dahlsailrunner** [comment]
Hi - Wondering if there's any new general "status" updates on v3 for general release.  I noticed that the last update here was about a year ago and was just wondering.  Thanks in advance.  (btw - I did see the updates on the [alpha-status page](https://xunit.net/docs/v3-alpha), but those are from July).

**Reply by bradwilson** [reply]
Recently I have been mostly focused on v2 releases as we've had a lot of outside contributions, plus I moved the rewritten assertion library from v3 over into v2 as well which has had a lot of knock-on effect of needing to provide fixes and additions. It is still my intention to get back to v3, but there has been little movement since July of last year when I did the v2 2.5.0 release.

**Reply by dahlsailrunner** [reply]
Sounds good and thanks for the update.

**Comment by mattiasnordqvist** [comment]
How's it going? Really looking forward to the Test Platform integration!

**Reply by bradwilson** [reply]
The majority of the v3 features that I wanted for 1.0 have been added, so v3 will definitely be out before the end of the year. I have a couple more extensibility things and some parallelism adjustments to work on, and then I'll want a basically feature complete pre-release out for a little while for usage just to make sure everything is working as expected.

If you want to try things out, CI builds are the best way: https://xunit.net/docs/using-ci-builds

There will likely be an 0.5.0 build pushed to NuGet in the next week or so. Since shipping `0.4.0-pre.20`, we've added the ability to create "external" serializers (for classes which cannot directly implement `IXunitSerializable`), added serializers for Guid and Uri, updated v3 for the new runner reporter support, and fixed a handful of issues (the biggest one being issues where source line information was sometimes missing in Test Explorer).
